src/receive_imf.rs

```diff
@@ -1654,40 +1654,27 @@ async fn apply_group_changes(
             false
         };
 
-    // Whether to allow any changes to the member list at all.
-    let allow_member_list_changes = if chat::is_contact_in_chat(context, chat_id, ContactId::SELF)
-        .await?
-        || self_added
-        || !mime_parser.has_chat_version()
-    {
-        // Reject old group changes.
-        chat_id
-            .update_timestamp(context, Param::MemberListTimestamp, sent_timestamp)
-            .await?
-    } else {
-        // Member list changes are not allowed if we're not in the group
-        // and are not explicitly added.
-        // This message comes from a Delta Chat that restored an old backup
-        // or the message is a MUA reply to an old message.
-        false
-    };
+    // Whether to allow any changes to the member list at all. Just reject old group changes.
+    let allow_member_list_changes = chat_id
+        .update_timestamp(context, Param::MemberListTimestamp, sent_timestamp)
+        .await?;
+
+    let is_from_in_chat = chat::is_contact_in_chat(context, chat_id, from_id).await?;
 
     // Whether to rebuild the member list from scratch.
-    let recreate_member_list = if allow_member_list_changes {
+    let recreate_member_list = allow_member_list_changes && {
         // Recreate member list if the message comes from a MUA as these messages do _not_ set add/remove headers.
-        // Always recreate membership list if self has been added.
-        if !mime_parser.has_chat_version() || self_added {
-            true
-        } else {
-            match mime_parser.get_header(HeaderDef::InReplyTo) {
+        (is_from_in_chat && !mime_parser.has_chat_version())
+            // Always recreate membership list if SELF has been added. The current/older versions of
+            // DC don't always set "In-Reply-To" to the latest message they sent (it depends on
+            // timings), so we have this heuristic here.
+            || self_added
```

Comment:
So, i need to bring back this check for SELF-addition, but it's the only place where this heuristic is needed. If `In-Reply-To` will always point to the latest message, this check could be removed someday.